### PR TITLE
Java: Sqlite connector

### PR DESCRIPTION
**Pull request by milderhc**
### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#dev-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: java/semantickernel-connectors-parent/semantickernel-connectors/pom.xml [hunk 1]
```diff
@@ -60,6 +60,33 @@
...
+            <groupId>org.xerial</groupId>
+            <artifactId>sqlite-jdbc</artifactId>
+            <version>3.42.0.0</version>
+        </dependency>
+
+
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+            <version>2.20.0</version>
```
Comment (johnoliver) [inline java/semantickernel-connectors-parent/semantickernel-connectors/pom.xml:74]: shouldnt need a version, should be set in the parent

File: java/semantickernel-connectors-parent/semantickernel-connectors/pom.xml [hunk 2]
```diff
@@ -60,6 +60,33 @@
             <groupId>org.junit.jupiter</groupId>
             <artifactId>junit-jupiter</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>org.xerial</groupId>
+            <artifactId>sqlite-jdbc</artifactId>
+            <version>3.42.0.0</version>
```
Comment (johnoliver) [inline java/semantickernel-connectors-parent/semantickernel-connectors/pom.xml:67]: can put this in the bom

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+                                while (resultSet.next()) {
+                                    String key = resultSet.getString("key");
+                                    String metadata = resultSet.getString("metadata");
+                                    String embedding = resultSet.getString("embedding");
+                                    String timestamp = resultSet.getString("timestamp");
+                                    entries.add(
+                                            new DatabaseEntry(key, metadata, embedding, timestamp));
+                                }
+                            } catch (SQLException e) {
+                                throw new RuntimeException("Failed to read all entries", e);
```
Comment (johnoliver) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java:206]: should be Mono.error, and something more specific like creating a StorageConnectorException

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/memory/MemoryRecordMetadata.java [hunk 4]
```diff
@@ -79,13 +87,14 @@ public String getAdditionalMetadata() {
...
      */
+    @JsonCreator
     public MemoryRecordMetadata(
-            boolean isReference,
-            @Nonnull String id,
-            @Nonnull String text,
-            @Nonnull String description,
-            @Nonnull String externalSourceName,
-            @Nonnull String additionalMetadata) {
+            @JsonProperty("reference") boolean isReference,
```
Comment (dsgrieve) [inline java/semantickernel-api/src/main/java/com/microsoft/semantickernel/memory/MemoryRecordMetadata.java:92]: These need to match what's in dotnet/src/SemanticKernel.Abstractions/Memory/MemoryRecordMetadata.cs

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Database {
+
+    static class DatabaseEntry {
```
Comment (dsgrieve) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java:13]: I think dotnet are wrong to have each connector have it's own DatabaseEntry. There is a DataEntryBase that has the key and timestamp. We might want to create a VectorDataEntry that sits along side it that has embedding and metadata in it and use that here (and in the others).

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/SqliteMemoryStore.java [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+public class SqliteMemoryStore implements MemoryStore {
+
+    private final Database dbConnector;
+    private Connection dbConnection;
+
+    public SqliteMemoryStore() {
+        this.dbConnector = new Database();
+    }
+
+    public Mono<Void> connectAsync(String filename) throws SQLException {
```
Comment (johnoliver) [inline java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/SqliteMemoryStore.java:32]: should probably have this called in the builder so the store is connected to a db when created

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java (hunk 5 above)
Comment (dsgrieve) [reply java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java:13]: For later.

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        }
+
+        public String getTimestamp() {
+            return timestamp;
+        }
+    }
+
+    private static final String TABLE_NAME = "SKMemoryTable";
+
+    public Mono<Void> createTableAsync(Connection connection) {
```
[the remark this answers is not in the document]
Comment (dsgrieve) [reply java/semantickernel-connectors-parent/semantickernel-connectors/src/main/java/com/microsoft/semantickernel/connectors/memory/sqlite/Database.java:45]: For later

**Review by dsgrieve — APPROVED** [review]
Will follow on with another PR to address some review comments that were not resolved in this PR.

**Pull request merged**